src/main/java/redis/clients/jedis/ClientSideCache.java

```diff
@@ -1,71 +1,108 @@
 
-  private final Map<ByteBuffer, Object> cache;
+  private final Map<ByteBuffer, Set<Long>> keyHashes;
+  private final ReentrantReadWriteLock rwl = new ReentrantReadWriteLock();
+  private final Lock readLock = rwl.readLock();
+  private final Lock writeLock = rwl.writeLock();
 
-  public ClientSideCache() {
-    this.cache = new HashMap<>();
+  protected ClientSideCache() {
+    this.keyHashes = new ConcurrentHashMap<>();
   }
 
-  /**
-   * For testing purpose only.
-   * @param map 
-   */
-  ClientSideCache(Map<ByteBuffer, Object> map) {
-    this.cache = map;
+  protected ClientSideCache(Map<ByteBuffer, Set<Long>> keyHashes) {
+    this.keyHashes = keyHashes;
   }
 
-  public final void clear() {
-    cache.clear();
-  }
+  public abstract void invalidateAll();
+
+  protected abstract void invalidateAll(Iterable<Long> hashes);
+
+  protected abstract void put(long hash, Object value);
 
-  public final void invalidateKeys(List list) {
+  protected abstract Object get(long hash);
+
+  final void invalidateKeys(List list) {
     if (list == null) {
-      clear();
+      invalidateAll();
       return;
     }
 
-    list.forEach(this::invalidateKey);
+    Set<Long> hashes = new HashSet<>();
+    list.forEach(key -> hashes.addAll(getHashes(key)));
+    invalidateAll(hashes);
   }
 
-  private void invalidateKey(Object key) {
-    if (key instanceof byte[]) {
-      cache.remove(convertKey((byte[]) key));
-    } else {
-      throw new JedisException("" + key.getClass().getSimpleName() + " is not supported. Value: " + String.valueOf(key));
+  private Set<Long> getHashes(Object key) {
+    if (!(key instanceof byte[])) {
+      throw new AssertionError("" + key.getClass().getSimpleName() + " is not supported. Value: " + String.valueOf(key));
     }
-  }
 
-  protected void setKey(Object key, Object value) {
-    cache.put(getMapKey(key), value);
+    final ByteBuffer mapKey = makeKey((byte[]) key);
+    readLock.lock();
+    try {
+      Set<Long> hashes = keyHashes.get(mapKey);
+      return hashes != null ? hashes : Collections.emptySet();
+    } finally {
+      readLock.unlock();
+    }
   }
 
-  protected <T> T getValue(Object key) {
-    return (T) getMapValue(key);
+  final <T> T getValue(Function<CommandObject<T>, T> loader, CommandObject<T> command, String... keys) {
+
+    final long hash = getHash(command);
+
+    T value = (T) get(hash);
+    if (value != null) {
+      return value;
+    }
+
+    value = loader.apply(command);
+    if (value != null) {
+      writeLock.lock();
+      try {
+        put(hash, value);
+        for (String key : keys) {
+          ByteBuffer mapKey = makeKey(key);
+          if (keyHashes.containsKey(mapKey)) {
+            keyHashes.get(mapKey).add(hash);
+          } else {
+            Set<Long> set = new HashSet<>();
+            set.add(hash);
+            keyHashes.put(mapKey, set);
+          }
+        }
+      } finally {
+        writeLock.unlock();
+      }
+    }
```

Comment:
What if there are cached values for this key? Should we remove them, given that the key was removed on the server, most probably? Should there be an `else` for the `if (value != null)`?

Reply:
Do you mean, what if there are cached values for this "command"? In that case we have already returned the result in Line#70.

Reply:
It seems redundant as it should be covered by `void invalidate(List list)` (previously `void invalidateKeys(List list)`).